Components/Core/Timer.cpp

```diff
@@ -1,34 +1,226 @@
 	// We make a timer named "Timer" with a callback function that does nothing, Autoreload false, and the default period of 1s.
 	// The timer ID is specified as (void *)this to provide a unique ID for each timer object - however this is not necessary for polling timers.
 	// The timer is created in the dormant state.
-	rtTimerHandle = xTimerCreate("Timer", DEFAULT_TIMER_PERIOD, pdFALSE, (void *)this, empty_callback);
+	rtTimerHandle = xTimerCreate("Timer", DEFAULT_TIMER_PERIOD, pdFALSE, (void *)this, CallbackFunction);
+	SOAR_ASSERT(rtTimerHandle, "Error Occurred, Timer not created");
+	timerState = UNINITIALIZED;
+}
+
+/**
+ * Constructor for callback enabled timer
+ * ! User has to add "Timer::CallbackFunction(rtTimerHandle);" in the callback function for accurate functioning of Timer States
+ * Default behaviour : ->Autoreload is set to false (One shot Timer)
+ * 					   ->Timer Period is 1000ms
+ * 					   ->The Callback function will be provided by the user while making sure to follow instruction above
+*/
+Timer::Timer(void (*TimerCallbackFunction_t)( TimerHandle_t xTimer ))
+{
+	rtTimerHandle = xTimerCreate("Timer", DEFAULT_TIMER_PERIOD, pdFALSE, (void *)this, TimerCallbackFunction_t);
+	SOAR_ASSERT(rtTimerHandle, "Error Occurred, Timer not created");
+	timerState = UNINITIALIZED;
+}
+
+/**
+ * @brief Default de-constructor makes a timer that can only be polled for state
+*/
+Timer::~Timer()
+{
+	if (xTimerDelete(rtTimerHandle, DEFAULT_TIMER_COMMAND_WAIT_PERIOD*2) == pdPASS) {
+		SOAR_PRINT("Timer has been deleted \n\n");
+	}
+}
+
+/**
+ * @brief Callback function for polling timers
+*/
+void Timer::CallbackFunction(TimerHandle_t xTimer){
+	Timer* ptrTimer = (Timer*)pvTimerGetTimerID(xTimer);
+	ptrTimer->timerState = COMPLETE;
 }
 
 /**
- * @brief Changes this timer object's RTOS timer period, returns true on success, returns false on failure (timer command queue full)
+ * @brief Changes timer period, Sets timer state back to uninitialized and stops timer
  */
 bool Timer::ChangePeriod(const uint32_t period_ms)
 {
-	if (xTimerChangePeriod(rtTimerHandle, MS_TO_TICKS(period_ms), DEFAULT_TIMER_COMMAND_WAIT_PERIOD) == pdTRUE)
+	if (xTimerChangePeriod(rtTimerHandle, MS_TO_TICKS(period_ms), DEFAULT_TIMER_COMMAND_WAIT_PERIOD) == pdTRUE) {
+		if (xTimerStop(rtTimerHandle, DEFAULT_TIMER_COMMAND_WAIT_PERIOD) == pdPASS) {
+			timerState = UNINITIALIZED;
+			return true;
+		}
+	}
+	return false;
+}
+
+/**
+ * @brief Changes timer period, Sets timer state back to counting and starts timer
+ */
+bool Timer::ChangePeriodAndStart(const uint32_t period_ms)
+{
+	if (xTimerChangePeriod(rtTimerHandle, MS_TO_TICKS(period_ms), DEFAULT_TIMER_COMMAND_WAIT_PERIOD) == pdTRUE) {
+		timerState = COUNTING;
+		return true;
+	}
+	return false;
+}
+
+/**
+ * @brief Starts the timer
+*/
+bool Timer::Start()
+{
+	if ((timerState == COMPLETE) || (timerState == COUNTING)) {
+		return false;
+	}
+	else if (timerState == PAUSED) {
+		ChangePeriod(remainingTimeBetweenPauses);
```

Comment:
If this is ChangePeriodAndStart, then we shouldn't also need to xTimerStart (Right now it will basically Change period (start), stop, then start again, it seems)